src/transformation/mesh_intersection/mesh_intersection.rs

```diff
@@ -256,380 +355,358 @@ fn extract_connected_components(
-) -> Point<Real> {
-    let base_id2 = mesh1.vertices().len() as u32;
-    let base_id12 = (mesh1.vertices().len() + mesh2.vertices().len()) as u32;
-
-    if vid < base_id2 {
-        mesh1.vertices()[vid as usize]
-    } else if vid < base_id12 {
-        pos12 * mesh2.vertices()[(vid - base_id2) as usize]
-    } else {
-        new_vertices12[(vid - base_id12) as usize]
+fn planar_gram_schmidt(v1: Vector3<f64>, v2: Vector3<f64>) -> (Vector3<f64>, Vector3<f64>) {
+    let u1 = v1;
+    let u2 = v2 - (v2.dot(&u1) / u1.norm_squared()) * u1;
+
+    let e1 = u1.normalize();
+    let e2 = u2.normalize();
+
+    (e1, e2)
+}
+
+fn project_point_to_segment(point: &Vector3<f64>, segment: &[Vector3<f64>; 2]) -> Vector3<f64> {
+    let dir = segment[1] - segment[0];
+    let local = point - segment[0];
+
+    let norm = dir.norm();
+    // restrict the result to the segment portion of the line.
+    let coeff = (dir.dot(&local) / norm).clamp(0., norm);
+
+    segment[0] + coeff * dir.normalize()
+}
+
+/// No matter how smart we are about computing intersections. It is always possible
+/// to create ultra thin triangles when a point lies on an edge of a tirangle. These
+/// are degenerate and need to be terminated with extreme prejudice.
+fn is_triangle_degenerate(
+    triangle: &[Vector3<f64>; 3],
+    epsilon_degrees: f64,
+    epsilon_distance: f64,
+) -> bool {
+    if smallest_angle(triangle) < epsilon_degrees {
+        return true;
+    }
+
+    let mut shortest_side = f64::MAX;
+    for i in 0..3 {
+        let p1 = triangle[i];
+        let p2 = triangle[(i + 1) % 3];
+
+        shortest_side = shortest_side.min((p1 - p2).norm());
     }
+
+    let mut worse_projection_distance = f64::MAX;
+    for i in 0..3 {
+        let dir = triangle[(i + 1) % 3] - triangle[(i + 2) % 3];
+        if dir.norm() < epsilon_distance {
+            return true;
+        }
+
+        let dir = dir.normalize();
+        let proj = (triangle[i] - triangle[(i + 2) % 3]).dot(&dir) * dir + triangle[(i + 2) % 3];
+
+        worse_projection_distance = worse_projection_distance.min((proj - triangle[i]).norm());
+    }
+
+    if worse_projection_distance < epsilon_distance {
+        return true;
+    }
+
+    false
 }
 
-fn extract_result(
+fn merge_triangle_sets(
+    mesh1: &GenericTriMesh<DefaultStorage>,
+    mesh2: &GenericTriMesh<DefaultStorage>,
+    triangle_constraints: &BTreeMap<&u32, Vec<[Point3<f64>; 2]>>,
     pos12: &Isometry<Real>,
-    mesh1: &TriMesh,
-    flip1: bool,
-    mesh2: &TriMesh,
     flip2: bool,
-    spade_handle_to_intersection: &[HashMap<(u32, FixedVertexHandle), (FeatureId, FeatureId)>; 2],
-    intersection_points: &HashMap<(FeatureId, FeatureId), [Point<Real>; 2]>,
-    triangulations1: &HashMap<u32, Triangulation>,
-    triangulations2: &HashMap<u32, Triangulation>,
-    new_vertices12: &mut Vec<Point<Real>>,
-    new_indices12: &mut Vec<[u32; 3]>,
+    metadata: &MeshIntersectionMetadata,
+    mut point_set: &mut RTree<TreePoint>,
```

Comment:
Is the `mut` necessary?

Reply:
It is, otherwise the code does not compile on my local.